File: backend/java/elo-calculator-backend/src/main/java/de/iits/elocalculatorbackend/player/controller/PlayerController.java
package de.iits.elocalculatorbackend.player.controller;

import de.iits.elocalculatorbackend.player.model.resource.PlayerResponseResource;
import de.iits.elocalculatorbackend.player.model.resource.PlayerUpsertResource;
import de.iits.elocalculatorbackend.player.service.PlayerService;
import jakarta.validation.Valid;
import lombok.AllArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.server.ResponseStatusException;

import java.util.List;

@RestController
@AllArgsConstructor
public class PlayerController {

    private PlayerService playerService;

    @GetMapping("/players")
    public ResponseEntity<List<PlayerResponseResource>> getAllPlayers() {
        return ResponseEntity.ok(playerService.findAll());
    }

    @PutMapping("/players")
    public ResponseEntity<PlayerResponseResource> updatePlayer(@RequestBody @Valid PlayerUpsertResource playerUpsertResource) {
        if (!playerService.existsById(playerUpsertResource.id())) {
            return ResponseEntity.notFound().build();
        }
        PlayerResponseResource updatedPlayer = playerService.updatePlayer(playerUpsertResource);
        return ResponseEntity.ok(updatedPlayer);
    }

    @PostMapping("/players")
    public ResponseEntity<PlayerResponseResource> createNewPlayer(@RequestBody @Valid PlayerUpsertResource playerUpsertResource) {
        return ResponseEntity.ok(playerService.createPlayer(playerUpsertResource));
    }

    @GetMapping("/players/{username}")
    public ResponseEntity<PlayerResponseResource> getPlayerByUsername(@PathVariable String username) {
        PlayerResponseResource player = playerService.findByUsername(username)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND));

        return ResponseEntity.ok(player);
    }

    @GetMapping("/players/{username}/elo")
    public ResponseEntity<Integer> getEloByUsername(@PathVariable String username) {
        PlayerResponseResource player = playerService.findByUsername(username)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND));

        return ResponseEntity.ok(player.elo());
    }
}

File: backend/springboot-kotlin/src/main/kotlin/de/iits/elo/player/PlayerController.kt
package de.iits.elo.player

import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.server.ResponseStatusException

@RestController
class PlayerController(
    private val playerService: PlayerService,
) {

    @GetMapping("/players")
    fun getAllPlayers(): ResponseEntity<List<Player>> =
        ResponseEntity.ok(playerService.findAll())

    @PutMapping("/players")
    fun updatePlayer(@RequestBody player: Player?): ResponseEntity<Player> {
        player ?: throw ResponseStatusException(HttpStatus.BAD_REQUEST, "Player required for update, but no player was found in request body")
        val update = try {
            playerService.update(player)
        } catch (exception: IllegalStateException) {
            throw ResponseStatusException(HttpStatus.NOT_FOUND, exception.message)
        }
        return ResponseEntity.ok(update)
    }

    @PostMapping("/players")
    fun createNewPlayer(@RequestBody player: Player?): ResponseEntity<Player> {
        player ?: throw ResponseStatusException(HttpStatus.BAD_REQUEST, "Player required for creation, but no player was found in request body")
        return ResponseEntity.ok(playerService.save(player))
    }

    @GetMapping("/players/{userName}")
    fun getPlayerByUserName(@PathVariable userName: String): ResponseEntity<Player> {
        val requestedUser = playerService.findByUserName(userName)
            ?: throw ResponseStatusException(HttpStatus.NOT_FOUND, "Could not find player with user name $userName")
        return ResponseEntity.ok(requestedUser)
    }

    @GetMapping("/players/{userName}/elo")
    fun getPlayerElo(@PathVariable userName: String): ResponseEntity<Int> =
        try {
            ResponseEntity.ok(playerService.getElo(userName))
        } catch (exception: IllegalStateException) {
            throw ResponseStatusException(HttpStatus.NOT_FOUND, exception.message)
        }
}
